bigquery/google/cloud/bigquery/dbapi/cursor.py

```diff
@@ -0,0 +1,280 @@
+        the output will be a query like ``SELECT @namedparam``.
+
+        :type: str
+        :param operation: A Google BigQuery query string.
+
+        :type: Mapping[str, Any]
+        :param parameters: Dictionary of parameter values.
+        """
+        formatted_params = {}
+        for name in parameters:
+            formatted_params[name] = '@{}'.format(name)
+
+        try:
+            return operation % formatted_params
+        except KeyError as ex:
+            raise exceptions.ProgrammingError(ex)
+
+    def _format_operation(self, operation, parameters=None):
+        """Formats parameters in operation in way BigQuery expects.
+
+        Raises a ProgrammingError if a parameter used in the operation is not
+        found in the parameters dictionary.
+
+        :type: str
+        :param operation: A Google BigQuery query string.
+
+        :type: Mapping[str, Any] or Sequence[Any]
+        :param parameters: Optional parameter values.
+        """
+        if parameters is None:
+            return operation
+
+        if isinstance(parameters, collections.Mapping):
+            return self._format_operation_dict(operation, parameters)
+
+        return self._format_operation_list(operation, parameters)
+
+    def execute(self, operation, parameters=None):
+        """Prepare and execute a database operation.
+
+        :type: str
+        :param operation: A Google BigQuery query string.
+
+        :type: Mapping[str, Any] or Sequence[Any]
+        :param parameters: Optional dictionary or sequence of parameter values.
+        """
+        self._query_results = None
+        self._page_token = None
+        self._has_fetched_all_rows = False
+        client = self.connection._client
+        job_id = str(uuid.uuid4())
+
+        # The DB-API uses the pyformat formatting, since the way BigQuery does
+        # query parameters was not one of the standard options. Convert both
+        # the query and the parameters to the format expected by the client
+        # libraries.
+        formatted_operation = self._format_operation(
+            operation, parameters=parameters)
+        query_parameters = _helpers.to_query_parameters(parameters)
+
+        query_job = client.run_async_query(
+            job_id,
+            formatted_operation,
+            query_parameters=query_parameters)
+        query_job.use_legacy_sql = False
+        query_job.begin()
+        _helpers.wait_for_job(query_job)
+        query_results = query_job.results()
+
+        # Force the iterator to run because the query_results doesn't
+        # have the total_rows populated. See:
+        # https://github.com/GoogleCloudPlatform/google-cloud-python/issues/3506
+        query_iterator = query_results.fetch_data()
+        try:
+            six.next(iter(query_iterator))
+        except StopIteration:
+            pass
+
+        self._query_data = iter(
+            query_results.fetch_data(max_results=self.arraysize))
```

Comment:
1. Can't you use this to set `total_rows`?
1. Is it not already an iterator? (i.e. why call `iter()`?)